From f9832072ca8096060fe39ded8aca13a8bc64de16 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Thu, 31 Dec 2020 05:24:30 -0500
Subject: [PATCH] win32: Use __MINGW_PRINTF_FORMAT instead of __printf__ for
 MinGW builds

---
 src/pulse/gccmacro.h | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/src/pulse/gccmacro.h b/src/pulse/gccmacro.h
index 81729dbba..4a0fff5f4 100644
--- a/src/pulse/gccmacro.h
+++ b/src/pulse/gccmacro.h
@@ -25,10 +25,8 @@
 
 #if defined(__GNUC__)
 #ifdef __MINGW32__
-/* libintl overrides printf with a #define. As this breaks this attribute,
- * it has a workaround. However the workaround isn't enabled for MINGW
- * builds (only cygwin) */
-#define PA_GCC_PRINTF_ATTR(a,b) __attribute__ ((format (__printf__, a, b)))
+#include <stdio.h>
+#define PA_GCC_PRINTF_ATTR(a,b) __attribute__ ((format (__MINGW_PRINTF_FORMAT, a, b)))
 #else
 #define PA_GCC_PRINTF_ATTR(a,b) __attribute__ ((format (printf, a, b)))
 #endif
-- 
2.29.2

