From 79f21cd712f976ca202cd3ce8eca329213d174a0 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Thu, 31 Dec 2020 06:13:06 -0500
Subject: [PATCH] win32: Fix privlibdir for running on Windows

---
 meson.build | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/meson.build b/meson.build
index cd617c602..888692e80 100644
--- a/meson.build
+++ b/meson.build
@@ -67,6 +67,11 @@ localstatedir = join_paths(prefix, get_option('localstatedir'))
 sysconfdir = join_paths(prefix, get_option('sysconfdir'))
 privlibdir = join_paths(libdir, 'pulseaudio')
 
+if host_machine.system() == 'windows'
+  # Windows only supports loading libraries from the same dir as the executable
+  privlibdir = bindir
+endif
+
 alsadatadir = get_option('alsadatadir')
 if alsadatadir == ''
   alsadatadir = join_paths(datadir, 'pulseaudio', 'alsa-mixer')
-- 
2.29.2

