From b4f9e6bdf23d4e0c240a2f58cf4a70efbdb8f254 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Sat, 2 Jan 2021 16:56:20 -0500
Subject: [PATCH] client: Make auto_connect_localhost respect HAVE_IPv6 and
 OS_IS_WIN32

---
 src/pulse/context.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/pulse/context.c b/src/pulse/context.c
index 05b6633aa..b0a838860 100644
--- a/src/pulse/context.c
+++ b/src/pulse/context.c
@@ -1027,7 +1027,10 @@ int pa_context_connect(
 
         /* Add TCP/IP on the localhost */
         if (c->conf->auto_connect_localhost) {
+#if defined(HAVE_IPV6) && !defined(OS_IS_WIN32)
+            /* FIXME: pa_socket_client does not support IPv6 on Windows */
             c->server_list = pa_strlist_prepend(c->server_list, "tcp6:[::1]");
+#endif
             c->server_list = pa_strlist_prepend(c->server_list, "tcp4:127.0.0.1");
         }
 
-- 
2.29.2

